lib/internal/modules/esm/loader.js

```diff
@@ -12,11 +12,16 @@ const defaultResolve = require('internal/modules/esm/default_resolve');
 const createDynamicModule = require(
   'internal/modules/esm/create_dynamic_module');
 const translators = require('internal/modules/esm/translators');
+const { SafeWeakSet } = require('internal/safe_globals');
 
 const FunctionBind = Function.call.bind(Function.prototype.bind);
 
 const debug = require('util').debuglog('esm');
 
+const thenableNamespaces = new SafeWeakSet();
+const thenableWarning = (x) => `The namespace of ${x} is thenable. \
+See https://mdn.io/thenable for more information.`;
```

Comment:
I'm not sure how obvious this is to users - perhaps `The module ${x} cannot be resolved through a dynamic import because it has a "then" method.`

Reply:
the first paragraph of the linked article explains it pretty well, do you feel strongly about adding more detail here?

Reply:
This should probably be updated now to reflect the new warning status - `The namespace of ${x} exports an invalid thenable, resulting in top-level imports of this module never resolving.`.